perf(core): introduce three shared thread pools - Query, Write and Network

ideoma opened the pull request:
This PR changes default QuestDB threading strategy. Instead of single shared pool there will be three shared thread pools:

- Network thread pool. This will run network IO e.g. HTTP, Pg and ILP servers.
- Query Thread pool. This pool will handle parallel queries, e.g. parallel filters and group by
- Writer Thread pool. This will run Wal Apply Job, Table Writer jobs, Mat View Refresh, various purge and housekeeping jobs.

tandem: https://github.com/questdb/questdb-enterprise/pull/645

glasstiger commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 145 / 153 (94.77%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/DefaultServerConfiguration.java|8|14|57.14%|
|:large_blue_circle:|io/questdb/cutlass/Services.java|6|7|85.71%|
|:large_blue_circle:|io/questdb/ServerMain.java|16|17|94.12%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|53|53|100.00%|
|:large_blue_circle:|io/questdb/WorkerPoolManager.java|52|52|100.00%|
|:large_blue_circle:|io/questdb/mp/WorkerPoolConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|6|6|100.00%|

glasstiger commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 145 / 153 (94.77%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/DefaultServerConfiguration.java|8|14|57.14%|
|:large_blue_circle:|io/questdb/cutlass/Services.java|6|7|85.71%|
|:large_blue_circle:|io/questdb/ServerMain.java|16|17|94.12%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|53|53|100.00%|
|:large_blue_circle:|io/questdb/WorkerPoolManager.java|52|52|100.00%|
|:large_blue_circle:|io/questdb/mp/WorkerPoolConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|6|6|100.00%|

bluestreak01 commented on core/src/main/java/io/questdb/cutlass/Services.java:
```diff
@@ -153,11 +153,11 @@ public LineTcpReceiver createLineTcpReceiver(
                 Requester.LINE_TCP_IO
         );
-        final WorkerPool writerPool = workerPoolManager.getInstance(
+        final WorkerPool writerPool = workerPoolManager.getInstanceWrite(
```
naming, it does not immediately jump at me what the difference between "IO" pool and "Write" pool. It reads that one is a subset of the other

bluestreak01 commented:
can i suggest to rename pools:
- Network Worker Pool
- Read IO Worker Pool
- Write IO Worker Pool

Subjectively this is clearer to me.

bluestreak01 commented on core/src/main/java/io/questdb/WorkerPoolManager.java:
```diff
@@ -34,55 +34,46 @@
 
-    public WorkerPool getSharedPool() {
-        return sharedPool;
+    public WorkerPool getSharedPoolNetwork() {
```
these methods should be renamed as we lost concept of "shared" pool

ideoma replied:
All three are still "shared" pools, as the opposite to "dedicated" pools.
Dedicated pools are still possible, you can turn them on for Pg, ILP, Wal Apply etc.

bluestreak01 commented on core/src/main/java/io/questdb/WorkerPoolManager.java:
```diff
@@ -34,55 +34,46 @@
+        return sharedPoolNetwork;
     }
 
     public int getSharedWorkerCount() {
```
this will now lead to potential bugs, not covered by tests. Such as passing the number of workers in the network pool to WAL apply job that has nothing to do with the network pool

bluestreak01 commented on core/src/main/java/io/questdb/ServerMain.java:
```diff
@@ -328,47 +328,47 @@ private synchronized void initialize() {
 
                         if (matViewEnabled && !config.getMatViewRefreshPoolConfiguration().isEnabled()) {
-                            setupMatViewJobs(sharedPool, engine, sharedPool.getWorkerCount());
+                            setupMatViewJobs(sharedPoolWrite, engine, sharedPoolWrite.getWorkerCount());
```
in fact this is just a bug - 3rd arg is used by mat view infra to execute queries and partition query data structures. What is passed from here is wrong

ideoma replied:
I renamed the parameter to `sharedQueryWorkerCount` and passing `sharedPoolQuery.getWorkerCount()` or `workerPoolManager.getSharedQueryWorkerCount()`

bluestreak01 reviewed (changes requested):
a lot more work is required here:

- parameter and field names everywhere need to be reviewed now that "shared pool" is not more
- there is proliferated confusion as to what "worker" count is meant to represent, wrong counts being passed around and parameter names are not helping
- we probably need some sort of fuzz test that boostraps unequal size pools and executes parallel queries on them, for example writer pool could be 4 and query pool could be 10. SQL may partition its data structures for 4 workers and will have 10 workers executing, causing chaos
- i observed weird boostrap, on 12 core laptop, network pool = 12, query pool = 12, wirter pool = 18? why 18?

glasstiger commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 150 / 158 (94.94%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/DefaultServerConfiguration.java|8|14|57.14%|
|:large_blue_circle:|io/questdb/cutlass/Services.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/ServerMain.java|16|17|94.12%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|56|56|100.00%|
|:large_blue_circle:|io/questdb/WorkerPoolManager.java|52|52|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpReceiverConfigurationWrapper.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/mp/WorkerPoolConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|6|6|100.00%|

glasstiger commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 150 / 158 (94.94%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/DefaultServerConfiguration.java|8|14|57.14%|
|:large_blue_circle:|io/questdb/cutlass/Services.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/ServerMain.java|16|17|94.12%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|56|56|100.00%|
|:large_blue_circle:|io/questdb/WorkerPoolManager.java|52|52|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpReceiverConfigurationWrapper.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/mp/WorkerPoolConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|6|6|100.00%|

glasstiger commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 181 / 194 (93.30%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/DefaultServerConfiguration.java|8|14|57.14%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|14|18|77.78%|
|:large_blue_circle:|io/questdb/ServerMain.java|19|21|90.48%|
|:large_blue_circle:|io/questdb/cutlass/Services.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|53|53|100.00%|
|:large_blue_circle:|io/questdb/cairo/sql/async/PageFrameSequence.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/WorkerPoolManager.java|52|52|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpReceiverConfigurationWrapper.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/mp/WorkerPoolConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/mv/MatViewRefreshJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalApplySqlExecutionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContextImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/PageFrameRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/mv/MatViewRefreshSqlExecutionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/TextQueryProcessor.java|1|1|100.00%|

glasstiger commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 181 / 194 (93.30%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/DefaultServerConfiguration.java|8|14|57.14%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|14|18|77.78%|
|:large_blue_circle:|io/questdb/ServerMain.java|19|21|90.48%|
|:large_blue_circle:|io/questdb/cutlass/Services.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|53|53|100.00%|
|:large_blue_circle:|io/questdb/cairo/sql/async/PageFrameSequence.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/WorkerPoolManager.java|52|52|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpReceiverConfigurationWrapper.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/mp/WorkerPoolConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/mv/MatViewRefreshJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalApplySqlExecutionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContextImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/PageFrameRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/mv/MatViewRefreshSqlExecutionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/TextQueryProcessor.java|1|1|100.00%|

ideoma commented on core/src/main/java/io/questdb/ServerMain.java:
```diff
@@ -328,47 +328,47 @@ private synchronized void initialize() {
             @Override
-            protected void configureSharedPool(WorkerPool sharedPool) {
+            protected void configureSharedPool(final WorkerPool sharedPoolIO, final WorkerPool sharedPoolQuery, final WorkerPool sharedPoolWrite) {
                 try {
```
removed

glasstiger commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 197 / 210 (93.81%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/DefaultServerConfiguration.java|8|14|57.14%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|15|19|78.95%|
|:large_blue_circle:|io/questdb/ServerMain.java|19|21|90.48%|
|:large_blue_circle:|io/questdb/cutlass/Services.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|54|54|100.00%|
|:large_blue_circle:|io/questdb/cairo/sql/async/PageFrameSequence.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/WorkerPoolManager.java|52|52|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpReceiverConfigurationWrapper.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/mp/WorkerPoolConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/mv/MatViewRefreshJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CopyRequestJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalApplySqlExecutionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContextImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/test/TestMatchFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/PageFrameRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/mv/MatViewRefreshSqlExecutionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/TextQueryProcessor.java|1|1|100.00%|

glasstiger commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 197 / 210 (93.81%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/DefaultServerConfiguration.java|8|14|57.14%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|15|19|78.95%|
|:large_blue_circle:|io/questdb/ServerMain.java|19|21|90.48%|
|:large_blue_circle:|io/questdb/cutlass/Services.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|54|54|100.00%|
|:large_blue_circle:|io/questdb/cairo/sql/async/PageFrameSequence.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/WorkerPoolManager.java|52|52|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpReceiverConfigurationWrapper.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/mp/WorkerPoolConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/mv/MatViewRefreshJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CopyRequestJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalApplySqlExecutionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContextImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/test/TestMatchFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/PageFrameRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/mv/MatViewRefreshSqlExecutionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/TextQueryProcessor.java|1|1|100.00%|

ideoma commented:
Hey @bluestreak01, this PR is now ready for a second look

> parameter and field names everywhere need to be reviewed now that "shared pool" is no more
there is proliferated confusion as to what "worker" count is meant to represent, wrong counts being passed around and parameter names are not helping

I renamed parameters and fields everywhere to be `sharedQueryWorkerCount` instead of `workerCount` and `sharedWorkerCount` so it's more obvious what they are. I also removed the pattern where we passed both `workerCount` and `sharedWorkderCount`, almost all the places only need to know `sharedQueryWorkerCount`

> we probably need some sort of fuzz test that boostraps unequal size pools and executes parallel queries on them, for example writer pool could be 4 and query pool could be 10. SQL may partition its data structures for 4 workers and will have 10 workers executing, causing chaos

I added test `UnequalWorkerPoolTest` where `SHARED_WORKER_COUNT=1` and `SHARED_QUERY_WORKER_COUNT=4`. The test runs parallelised queries.

> i observed weird boostrap, on 12 core laptop, network pool = 12, query pool = 12, wirter pool = 18? why 18?

I changed the writer pool to 12. It used to be 50% more than the other pool because it's not supposed to be CPU-bound, but I decided against it now to avoid oversaturating the IO.

glasstiger commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 197 / 210 (93.81%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/DefaultServerConfiguration.java|8|14|57.14%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|15|19|78.95%|
|:large_blue_circle:|io/questdb/ServerMain.java|19|21|90.48%|
|:large_blue_circle:|io/questdb/cutlass/Services.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|54|54|100.00%|
|:large_blue_circle:|io/questdb/cairo/sql/async/PageFrameSequence.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/WorkerPoolManager.java|52|52|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpReceiverConfigurationWrapper.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/mp/WorkerPoolConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/mv/MatViewRefreshJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CopyRequestJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalApplySqlExecutionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContextImpl.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/test/TestMatchFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/PageFrameRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/mv/MatViewRefreshSqlExecutionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/TextQueryProcessor.java|1|1|100.00%|
